feat(samples): mcp client with kb

adiadd opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #127 

## Summary

### Changes

Addition of a sample MCP Client and user interface interacting with the `bedrock-kb-retrieval-mcp-server` MCP server

### User experience

Users will be able to view an example implementation of integrating MCP with Bedrock, creating a bridge between the model's ability to request information (via the Converse API) and MCP's standardized protocol for accessing external systems.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented

> This is not a breaking change

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

scottschreckengaust commented:
Hi @adiadd, thank you for the contribution. We have pre-commit at the root of the repository running. You can review those by running:

```shell
% pre-commit run --all-files
```

Some issues will automatically fix and others appear [docstring](https://peps.python.org/pep-0257/) related. After passing those, commit those changes to check.

adiadd commented:
Hi @scottschreckengaust, have run the `pre-commit` command and all checks are passing. Appreciate it!

scottschreckengaust commented on README.md:
```diff
@@ -221,7 +225,7 @@ Documentation for each server:
 - [Cost Analysis MCP Server](https://awslabs.github.io/mcp/servers/cost-analysis-mcp-server/)
 - [Amazon Nova Canvas MCP Server](https://awslabs.github.io/mcp/servers/nova-canvas-mcp-server/)
-- [AWS Diagram MCP Server]((https://awslabs.github.io/mcp/servers/aws-diagram-mcp-server/))
+- [AWS Diagram MCP Server](<(https://awslabs.github.io/mcp/servers/aws-diagram-mcp-server/)>)
```
```suggestion
- [AWS Diagram MCP Server](https://awslabs.github.io/mcp/servers/aws-diagram-mcp-server/)
```

scottschreckengaust commented on samples/mcp-integration-with-kb/README.md:
```diff
@@ -0,0 +1,82 @@
+### Prerequisites
+
+- The [uv](https://docs.astral.sh/uv/getting-started/installation/) package manager
+- AWS Account with Bedrock access and proper IAM permissions
```
Consider a link to the [Getting started with Amazon Bedrock](https://docs.aws.amazon.com/bedrock/latest/userguide/getting-started.html)

scottschreckengaust commented on samples/mcp-integration-with-kb/README.md:
````diff
@@ -0,0 +1,82 @@
+1. Clone the repository.
+
+```bash
+git clone <INSERT_REPO_URL>
````
```suggestion
git clone https://github.com/awslabs/mcp.git
```

scottschreckengaust commented on samples/mcp-integration-with-kb/README.md:
````diff
@@ -0,0 +1,82 @@
+git clone <INSERT_REPO_URL>
+```
+
+2. Copy the .env.example file to .env and add your AWS credentials.
+
+```bash
+cp .env.example .env
+```
````
```suggestion
2. Got the sample directory and copy the .env.example file to .env and add your AWS credentials.

```bash
cd mcp/samples/mcp-integration-with-kb
cp .env.example .env
```
```

adiadd commented:
Great @scottschreckengaust ! Addressed comments and updated the PR 😄

krokoko commented on samples/mcp-integration-with-kb/clients/client_server.py:
```diff
@@ -0,0 +1,269 @@
+import boto3
```
missing license header in all python files

dineshSajwan commented on samples/mcp-integration-with-kb/.env.example:
```diff
@@ -0,0 +1,5 @@
+AWS_ACCESS_KEY_ID=<your-aws-access-key-id>
```
Nit pick. would suggest to pick these from aws cli profile or CI/CD env file then keeping it in env file. This is to avoid any accidental check in of secrets in the env file.

adiadd replied:
Agreed! We actually don't even need the specific AWS environment information and it automatically picks it up from AWS CLI profile - updated the `.env.example` and `client_server.py` files accordingly

adiadd commented:
Awesome, updated the PR addressing the comments @dineshSajwan @krokoko 😄!

krokoko commented:
> Awesome, updated the PR addressing the comments @dineshSajwan @krokoko 😄!

Thanks @adiadd ! I'm testing it right now, will post my comments right after, if all good it LGTM !

krokoko commented:
@adiadd Following the instructions step by step. When making a request against the KB, the following error appears:

```
2025-04-14 13:52:10,446 - clients.client_server - ERROR - Error in process_query: Error executing tool QueryKnowledgeBases: An error occurred (ValidationException) when calling the Retrieve operation: Your account is unauthorized to use RerankingConfiguration.
```

However the KB is properly configured as per the [documentation](https://awslabs.github.io/mcp/servers/bedrock-kb-retrieval-mcp-server/).
- Running everything is us-east-1
- My profile has Admin access
- My KB service role has AmazonBedrockFullAccess
<img width="302" alt="image" src="https://github.com/user-attachments/assets/ad683cc2-3ed8-4cf1-aad5-4309383a4b8b" />

Anything obvious I missed ? Probably need to update the README to provide a bit more information on the setup required

adiadd commented:
> Update: answering myself: documentation of the mcp server is incorrect, reranking is not available in us-east-1
> 
> @adiadd Following the instructions step by step. When making a request against the KB, the following error appears:
> 
> ```
> 2025-04-14 13:52:10,446 - clients.client_server - ERROR - Error in process_query: Error executing tool QueryKnowledgeBases: An error occurred (ValidationException) when calling the Retrieve operation: Your account is unauthorized to use RerankingConfiguration.
> ```
> 
> However the KB is properly configured as per the [documentation](https://awslabs.github.io/mcp/servers/bedrock-kb-retrieval-mcp-server/).
> 
>     * Running everything in region us-east-1
> 
>     * My profile has Admin access
> 
>     * My KB service role has AmazonBedrockFullAccess
> 
> 
> <img alt="image" width="302" src="https://private-user-images.githubusercontent.com/6213264/433520932-ad683cc2-3ed8-4cf1-aad5-4309383a4b8b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NjAwMjYsIm5iZiI6MTc0NDY1OTcyNiwicGF0aCI6Ii82MjEzMjY0LzQzMzUyMDkzMi1hZDY4M2NjMi0zZWQ4LTRjZjEtYWFkNS00MzA5MzgzYTRiOGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MTRUMTk0MjA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWNjZTkyOWM1NTBhOTAyNDFlODlkY2JkYTY0OWE3NDFiZGM3MjgyN2EwMzY3ZmVmZDcyM2ViYWQxMmRkMjQxZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.0O-5VGZHhYpllCsFGNnrViPcWsvx-uEpu36ofMh4zO4">
> 
> Anything obvious I missed ? Probably need to update the README to provide a bit more information on the setup required

Just saw your edit and yep, looks like [rerank isn't available](https://docs.aws.amazon.com/bedrock/latest/userguide/rerank-supported.html) in `us-east-1`. Would need to update the MCP server - I can also update the README for this sample + set the default backup region (if not specified in `.env`) to `us-west-2`. Please do let me know if anything else is not working properly or if there are any steps/clarifications to adjust in the README!

krokoko commented:
@adiadd Indeed the issue was the region. It works as expected when deploying in a supported region (here us-west-2). I opened this issue : https://github.com/awslabs/mcp/issues/140 to fix the [documentation](https://docs.aws.amazon.com/bedrock/latest/userguide/rerank-supported.html) of the server to access the up to date list of supported regions. I think in your sample you can just add a direct link to the AWS documentation specifically for reranking

<img width="1907" alt="Screenshot 2025-04-14 at 2 51 52 PM" src="https://github.com/user-attachments/assets/3a789444-5b84-4338-8cf7-9a394e27fc0e" />

Besides that no issue, great job !

adiadd commented:
Done! Added a note that links to the documentation and changed the backup region to `us-west-2` - please do let me know if there's anything else needed to get this merged 😄!
